src/ApplicationRecordLog.cpp

```diff
@@ -0,0 +1,257 @@
+                    .time = geopm_time_diff(&m_time_zero, &time),
+                    .process = m_process,
+                    .event = EVENT_REGION_ENTRY,
+                    .signal = hash,
+                };
+                enter_info.record_idx = layout.num_record;
+                append_record(layout, enter_record);
+            }
+            GEOPM_DEBUG_ASSERT(enter_info.record_idx >= 0 && enter_info.record_idx < layout.num_record,
+                               "Invalid record index");
+
+            // find or add the region in short regions array
+            int region_idx = enter_info.region_idx;
+            if (region_idx == -1) {
+                region_idx = layout.num_region;
+                enter_info.region_idx = region_idx;
+                ++(layout.num_region);
+                GEOPM_DEBUG_ASSERT(layout.num_region <= M_MAX_REGION,
+                                   "ApplicationRecordLogImp::exit(): too many regions entered and exited within one control loop");
+                layout.region_table[region_idx] = {
+                    .hash = hash,
+                    .num_complete = 0,
+                    .total_time = 0.0,
+                };
+                // Convert the region entry event into a short region event
+                layout.record_table[enter_info.record_idx].event = EVENT_SHORT_REGION;
+                layout.record_table[enter_info.record_idx].signal = region_idx;
+            }
+            GEOPM_DEBUG_ASSERT(region_idx >= 0 && region_idx < layout.num_region,
+                               "Invalid region index");
+            auto &region = layout.region_table[region_idx];
+            ++(region.num_complete);
+            region.total_time += geopm_time_diff(&(enter_info.enter_time), &time);
+        }
+        m_entered_region_hash = GEOPM_REGION_HASH_INVALID;
+    }
+
+    void ApplicationRecordLogImp::epoch(const geopm_time_s &time)
+    {
+        check_setup();
+        std::unique_ptr<SharedMemoryScopedLock> lock = m_shmem->get_scoped_lock();
+        m_layout_s &layout = *((m_layout_s *)(m_shmem->pointer()));
+        check_reset(layout);
+
+        ++m_epoch_count;
+        record_s epoch_record = {
+           .time = geopm_time_diff(&m_time_zero, &time),
+           .process = m_process,
+           .event = EVENT_EPOCH_COUNT,
+           .signal = m_epoch_count,
+        };
+        append_record(layout, epoch_record);
+    }
+
+    void ApplicationRecordLogImp::dump(std::vector<record_s> &records,
+                                       std::vector<short_region_s> &short_regions)
+    {
+        // this function should not do anything with m_hash_region_enter_map
+        std::unique_ptr<SharedMemoryScopedLock> lock = m_shmem->get_scoped_lock();
+        m_layout_s &layout = *((m_layout_s *)(m_shmem->pointer()));
+        records.resize(layout.num_record);
+        std::copy(layout.record_table, layout.record_table + layout.num_record, records.begin());
```

Comment:
I think this calls the constructors of each entry twice (first default-constructed, then copy-constructed). The duplicate construction could be avoided with `records.assign(layout.record_table, layout.record_table + layout.num_record)`. I assume there is a pointer specialization with a built-in reserve call to avoid unnecessary reallocations, but you could precede it with `reserve` if necessary.